### PR TITLE
Generate pages with route parameters

### PR DESCRIPTION
**Pull request by Tobbe**
This PR adds support to the page generator for generating pages with route parameters.

Let's say you want to generate a page for displaying posts. You probably already know from the start that you will have to take the post ID as a route parameter. Now you can directly generate a page for that.

Is this useful? How do you like the syntax? What about the naming of everything? 

The new generator cli syntax looks like this
```bash
$ yarn rw g page post {id}
```

That will give you an entry in your routes like this:
```jsx
<Route path="/post/{id}" page={PostPage} name="post" />
```

and the generated page will look like this:
```jsx
import { Link } from '@redwoodjs/router'

const PostPage = ({ id }) => {
  return (
    <>
      <h1>PostPage</h1>
      <p>Find me in "./web/src/pages/PostPage/PostPage.js"</p>
      <p>
        My default route is named "post", link to me with `
        <Link to={routes.post({ id: 42 })}>Post 42</Link>`
      </p>
      <p>The parameter passed to me is {id}</p>
    </>
  )
}

export default PostPage
```

You'll also get a test that uses a parameter
```jsx
import { render } from '@redwoodjs/testing'

import PostPage from './PostPage'

describe('PostPage', () => {
  it('renders successfully', () => {
    expect(() => {
      render(<PostPage id={42} />)
    }).not.toThrow()
  })
})
```

Other supported generator cli syntaxes

```
$ yarn rw g page post /post/{id}
$ yarn rw g page post {id:Int}
$ yarn rw g page EditPost post/{id}/edit
$ yarn rw g page EditPost post/{id:Int}/edit
```

TODO:
- [x] Determine if this is something we want to add to RW
- [x] Actually launch/run a project that uses the generated pages
- [x] Run the generated tests
- [x] ~Make sure it handles spaces, like in `yarn rw g page post /post/{ id }`~ We don't want to support this (yargs gives a pretty useful error message if someone tries something like this)
- [x] Update docs https://github.com/redwoodjs/redwoodjs.com/issues/260
- [x] Make sure everything can be destroyed with the `destroy` cli command

### CODE REVIEW

**Comment by mojombo** [comment]
I like this! I think it makes total sense to add.

**Comment by Tobbe** [comment]
With the current implementation you get a 404 page if you go to /post. Should I add a route for that as well? So that the generator produces this:

```jsx
<Route path="/post" page={PostPage} name="post" />
<Route path="/post/{id}" page={PostPage} name="post" />
```

And then also maybe handle that case in the generated page

```jsx
import { Link } from '@redwoodjs/router'

const PostPage = ({ id }) => {
  return (
    <>
      <h1>PostPage</h1>
      <p>Find me in "./web/src/pages/PostPage/PostPage.js"</p>
      <p>
        My default route is named "post", link to me with `
        <Link to={routes.post({ id: 42 })}>Post 42</Link>`
      </p>
      <p>The parameter passed to me is {id ?? 'undefined'}</p>
    </>
  )
}

export default PostPage
```

(This is the updated line above: `<p>The parameter passed to me is {id ?? 'undefined'}</p>`)

**Comment by mojombo** [comment]
Hmmmm, good question. I'm leaning towards not creating the bare route because it could be surprising to create TWO routes for a single page generator, and then you'll essentially be required to check the params to ensure they're valid instead of letting it fall back to a 404.

Ok, I've decided: let's not include that extra route for now, it's easy enough to add the route entry manually if you want the nicer "index" page, but in doing so you'll be opting in to the extra work necessary to make that page not blow up.

**Comment by mojombo** [comment]
This is looking great! @Tobbe would you be up for adding corresponding docs to the CLI documentation on https://github.com/redwoodjs/redwoodjs.com?

**Comment by Tobbe** [comment]
@mojombo https://github.com/redwoodjs/redwoodjs.com/pull/273

**Comment by Tobbe** [comment]
One thing to note here is that this doesn't support multiple route parameters. I do however think it degrades gracefully enough to merge it as it is, and then leave support for multiple route parameters for a future PR.

`$ yarn rw g page post /posts/{year}/{month}/{day}/{slug}` gives you a page that only takes `{year}` as its prop. `const PostPage = ({ year }) => {`. The Routes entry looks like this `<Route path="/posts/{year}/{month}/{day}/{slug}" page={PostPage} name="post" />`. So everything should still at least render. But it's up to the user to add the other parameters to the list of props in the component

**Comment by Tobbe** [comment]
@mojombo The docs are written, and the PR is approved. Just waiting for this to be merged before merging the docs. Anything else you want me to do on this PR?

File: packages/cli/src/commands/generate/page/templates/page.js.template [hunk 1]
```diff
@@ -1,14 +1,17 @@
-import { Link } from '@redwoodjs/router'
+import { Link, routes } from '@redwoodjs/router'
 
-const ${pascalName}Page = () => {
+const ${pascalName}Page = (<%= typeof param !== 'undefined' && param ? param.replace(/:[^}]+/, '') : '' %>) => {
```
Comment (peterp) [inline packages/cli/src/commands/generate/page/templates/page.js.template:3]: Is there any way you can move this logic outside of the template?

Comment (Tobbe) [reply packages/cli/src/commands/generate/page/templates/page.js.template:3]: This prompted me to do something I should have done from the start - write tests for the `handler` method and its `Listr` tasks.
I'm getting there, but not quite done yet. Will continue tomorrow

Comment (peterp) [reply packages/cli/src/commands/generate/page/templates/page.js.template:3]: Ooooh, that's some ambitious work!  Thanks Tobbe !

File: packages/cli/src/commands/generate/page/__tests__/page.test.js [hunk 2]
```diff
@@ -1,16 +1,56 @@
 global.__dirname = __dirname
+
+let mockFs = false
+let mockFiles = {}
+
+jest.mock('fs', () => {
```
Comment (Tobbe) [inline packages/cli/src/commands/generate/page/__tests__/page.test.js:6]: I did see https://github.com/redwoodjs/redwood/blob/main/packages/cli/__mocks__/fs.js but I needed more control over when the mocked implementation was used, and when the real implementation was used.

Should I keep this here, as it is, or should I add this to `__mocks__/fs.js`? Or is there some third option I should look in to?

File: packages/cli/src/lib/index.js [hunk 3]
```diff
@@ -326,7 +326,7 @@ export const addRoutesToRouterTask = (routes) => {
     if (content.includes(route)) {
       return content
     }
-    return content.replace(/(\s*)\<Router\>/, `$1<Router>$1  ${route}`)
+    return content.replace(/<Router>(\s*)/, `<Router>$1${route}$1`)
```
Comment (Tobbe) [inline packages/cli/src/lib/index.js:329]: Reading my own PR I finally understand how this is supposed to work. Will have to revert this change and update my test instead...

File: packages/cli/src/commands/generate/page/__tests__/page.test.js [hunk 4]
```diff
@@ -106,3 +162,97 @@ test('creates a path equal to passed path', () => {
...
+    path.normalize(
+      '/path/to/project/web/src/pages/PostPage/PostPage.stories.js'
+    ),
+    loadGeneratorFixture('page', 'paramPage.stories.js')
+  )
+
+  expect(spy).toHaveBeenCalledWith(
+    path.normalize('/path/to/project/web/src/Routes.js'),
+    [
+      '<Router>',
```
Comment (Tobbe) [inline packages/cli/src/commands/generate/page/__tests__/page.test.js:248]: This isn't close enough to what the actual Routes.js file would look like. (Also see comment further down, with the regex stuff).

File: packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.stories.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import PostPage from './PostPage'
+
+export const generated = () => {
+  return <PostPage />
```
Comment (Tobbe) [inline packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.stories.js:4]: This should include the parameter.

Haven't used storybook before. Should I just hard code it? `id={42}`, or do I pass it to `generated = (id) => { //...`?

File: packages/cli/src/commands/generate/page/templates/page.js.template (hunk 1 above)
Comment (Tobbe) [reply packages/cli/src/commands/generate/page/templates/page.js.template:3]: Haven't fixed this yet, but I have the tests for `handler()` in place. Could you please take a look at how I test it? Is my approach sound? Also, if you could answer any of my questions I left as review comments on my own code that would be greatly appreciated 🙂

File: packages/cli/src/lib/index.js (hunk 3 above)
Comment (Tobbe) [reply packages/cli/src/lib/index.js:329]: Keeping this as it's a bit shorter, and a bit more flexible as well (with the indentation)

File: packages/cli/src/commands/generate/page/__tests__/page.test.js (hunk 4 above)
Comment (Tobbe) [reply packages/cli/src/commands/generate/page/__tests__/page.test.js]: Fixed

File: packages/cli/src/commands/generate/page/__tests__/page.test.js [hunk 6]
```diff
@@ -106,3 +162,97 @@ test('creates a path equal to passed path', () => {
...
+    path.normalize(
+      '/path/to/project/web/src/pages/HomePage/HomePage.stories.js'
+    ),
+    loadGeneratorFixture('page', 'singleWordPage.stories.js')
+  )
+
+  expect(spy).toHaveBeenCalledWith(
+    path.normalize('/path/to/project/web/src/Routes.js'),
+    [
+      '<Router>',
```
[the remark this answers is not in the document]
Comment (Tobbe) [reply packages/cli/src/commands/generate/page/__tests__/page.test.js]: Updated to include more realistic file contents

File: packages/cli/src/commands/generate/page/templates/page.js.template (hunk 1 above)
Comment (Tobbe) [reply packages/cli/src/commands/generate/page/templates/page.js.template]: I've moved the logic to a helper function in `page.js` that populates a bunch of template variables

File: packages/cli/src/commands/generate/page/__tests__/page.test.js (hunk 2 above)
Comment (peterp) [reply packages/cli/src/commands/generate/page/__tests__/page.test.js:6]: I think it's fine to keep this here.

File: packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { Link, routes } from '@redwoodjs/router'
+
+const PostPage = ({ id }) => {
+  return (
+    <>
+      <h1>PostPage</h1>
+      <p>Find me in "./web/src/pages/PostPage/PostPage.js"</p>
+      <p>
+        My default route is named "post", link to me with `
+        <Link to={routes.post({ id: '42' })}>Post 42</Link>`
```
Comment (peterp) [inline packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.js:10]: Would a paramPage always have an `id`?

File: packages/cli/src/commands/generate/page/templates/page.js.template (hunk 1 above)
Comment (peterp) [reply packages/cli/src/commands/generate/page/templates/page.js.template]: Amazing!

File: packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.js (hunk 7 above)
Comment (Tobbe) [reply packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.js:10]: I'm not sure I understand what you're getting at.
A paramPage will always have a parameter, but the name is whatever the user wants. For the test that corresponds with this fixture the name of the param is `id`. It comes from https://github.com/redwoodjs/redwood/pull/882/files#diff-6bcf3b847a12bd956cbf4aa88653206eR65

Comment (peterp) [reply packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.js:10]: Sorry, I was looking at this:
https://github.com/redwoodjs/redwood/pull/882/files#diff-55ff273ce0992ec67f307ed5ba69a541R30-R36

So, the generated param's value would always be `42`? And these values are only used in test and story templates?

Comment (Tobbe) [reply packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.js:10]: Yes, the example value will always be `"42"`. I was thinking of maybe making it `"exampleValue"`, or some other string for `{whatever}`, and `42`, for `{whatever:Int}`. But right now the generator isn't that clever. 

It's not used in the story template yet (see my other review comment/question and also the issue about Storybook on Windows). So right now the value is used in the page and test template, but *not* the story template.

Comment (peterp) [reply packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.js:10]: Ok, awesome! Thanks for the clarification, Maybe we could add a comment for the readers of the code?
`// the 42 values are example values that are used in tests and stories`

File: packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.stories.js (hunk 5 above)
Comment (peterp) [reply packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.stories.js:4]: I would include it in the same way you're doing it for the tests.

Comment (Tobbe) [reply packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.stories.js]: Done. But not tested, since Storybook isn't working for me

File: packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.js (hunk 7 above)
Comment (Tobbe) [reply packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.js:10]: Comment added. Please let me know if you want me to explain it further in the comment. https://github.com/redwoodjs/redwood/pull/882/commits/a90364b865cc477aecdc62ab5ef149a767687504

File: packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.stories.js (hunk 5 above)
Comment (peterp) [reply packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.stories.js]: Ok thanks, I saw the issue, but will take a closer look in the issue itself

File: packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.js (hunk 7 above)
Comment (peterp) [reply packages/cli/src/commands/generate/page/__tests__/fixtures/paramPage.js:10]: Thanks, that's perfect!
